### DISCUSSION TITLE
Allow `HttpServer::new`'s arg `factory` function to return a `Result`

### DISCUSSION THREAD

**Discussion by jayvdb** [actix/actix-web#3045] [Ideas]
We use functions to build the `App` like

```rust
pub fn build_app() -> App<
    impl ServiceFactory<
        ServiceRequest,
        Config = (),
        Response = ServiceResponse,
        Error = actix_web::Error,
        InitError = (),
    >,
> { ...}
```

However there is quite a bit of logic in these functions, with lots of bits that can fail, and we'd like to have structured errors for those, so something like

```rust
pub fn build_app() -> Result<App<
    impl ServiceFactory<
        ServiceRequest,
        Config = (),
        Response = ServiceResponse,
        Error = actix_web::Error,
        InitError = (),
    >>, Error
> { ... }
```

Is this possible already somehow?  If not, is it feasible to add?

The `InitError` there is very curious...

The way we work around this atm is the have a `fn setup_data(..) -> Result<AppData, ..>` and a `fn build_app(app_data: AppData) -> App<impl ServiceFactory<..>> { }` and then join them using

```rust
let data = setup_data(&env)?;

actix_web::HttpServer::new(move || build_app(data.clone()))....
```

**Comment by robjtede** [comment]
It's not currently possible and would require a breaking change; worth opening an issue imo.

There's some limited ways to do it for app data using `.data_factory()` but it's not quite what you're asking.

For now you just have to do all the fallible stuff before `HttpServer::new`.
